src/Extensions/ServiceFabricGuest.Abstractions/internal/ServiceFabricClientWrapper.cs

```diff
@@ -2,32 +2,68 @@
 // Licensed under the MIT license.
 
 using System;
+using System.Collections.Generic;
+using System.Security.Cryptography.X509Certificates;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Extensions.Options;
+using Microsoft.Omex.Extensions.Hosting.Certificates;
 using Microsoft.ServiceFabric.Client;
+using Microsoft.ServiceFabric.Common.Security;
 
 namespace Microsoft.Omex.Extensions.ServiceFabricGuest.Abstractions
 {
 	internal class ServiceFabricClientWrapper : IServiceFabricClientWrapper
 	{
 		private IServiceFabricClient? m_client;
 
-		private Uri m_clusterEndpoint;
+		private readonly ServiceFabricRestClientOptions m_options;
 
-		public ServiceFabricClientWrapper(IOptions<ServiceFabricRestClientOptions> options) => m_clusterEndpoint = new(options.Value.ClusterEndpoint());
+		private readonly ICertificateReader m_certificateReader;
+
+		public ServiceFabricClientWrapper(IOptions<ServiceFabricRestClientOptions> options, ICertificateReader certificateReader)
+		{
+			m_options = options.Value;
+			m_certificateReader = certificateReader;
+		}
 
 		public async Task<IServiceFabricClient> GetAsync(CancellationToken token)
 		{
 			if (m_client == null)
 			{
-				m_client = await new ServiceFabricClientBuilder()
-					.UseEndpoints(m_clusterEndpoint)
-					.BuildAsync(token)
-					.ConfigureAwait(false);
+				m_client = await CreateClientAsync(token).ConfigureAwait(false);
 			}
 
 			return m_client;
 		}
+
+		private Task<IServiceFabricClient> CreateClientAsync(CancellationToken token)
+		{
+			Uri clusterEndpoint = m_options.GetClusterEndpoint();
+
+			ServiceFabricClientBuilder builder = new ServiceFabricClientBuilder().UseEndpoints(clusterEndpoint);
+
+			if (clusterEndpoint.Scheme == Uri.UriSchemeHttps)
+			{
+				if (string.IsNullOrWhiteSpace(m_options.ClusterCertCommonName))
+				{
+					throw new InvalidOperationException($"{nameof(m_options.ClusterCertCommonName)} could not be empty for {Uri.UriSchemeHttps} cluster.");
+				}
+
+				X509Certificate2? clusterCert = m_certificateReader.GetCertificateByCommonName(m_options.ClusterCertCommonName);

```

Comment:
If there's multiple installed certificates matching the Common Name, which one is used?

Reply:
Logic is here https://github.com/microsoft/Omex/blob/3ec60e96b5068837302ff51579e2c1cb7ecdadbc/src/Extensions/Hosting/Certificates/CertificateReader.cs#L97

Reply:
I happened to be looking at that as well. I wonder why we even bother to return expired certificates?

Reply:
Not sure, initially I was trying to duplicate old behavior, but thinking about it I don't see a good reason to have an expired cert, unless we are getting better error message when it's expired compared to when it's just missing.

@aszczepanski maybe you know why are we returning expired certs?